docs/developer_guide/unit_test_design_guidelines.md

```diff
@@ -53,61 +61,82 @@ def test_feature_transformer(
 ```
 
-- When testing a complex layer / module / encoder / combiner / model, make sure that all the variables / weights get updates after one training step.  **Rationale**: Ensures the computation graph doesn’t contain dangling nodes. This catches issues that don’t make the code crash, that are not caught by looking at the loss scores, are they likely will go down, and that are not caught by training the model to convergence (albeit to a usually bad loss), For more details see [this link](https://thenerdstation.medium.com/how-to-unit-test-machine-learning-code-57cf6fd81765).
+### Trainable Modules
+
+When testing a trainable module (layer, encoder, decoder, combiner or model), make sure that all the variables / weights
+get updates after one training step. This will ensure that the computation graph does not contain dangling nodes. This
+catches subtle issues which don’t manifest as crashes, which are not caught by looking at the loss scores or by training
+the model to convergence (albeit to a usually bad loss), For more details see
+[how to unit test machine learning code](https://thenerdstation.medium.com/how-to-unit-test-machine-learning-code-57cf6fd81765).
+
+Add **type checking** based on torch input and outputs. Ensure all module outputs have the expected torch datatype,
+dimensionality, and tensor shape.
+
+For trainable modules, we recommend adding at least one **overfitting test**. Ensure that a small ECD model containing
+the module is able to overfit on a small dataset. Ensures that models are able to converge on reasonable targets and
+catches any unscoped issues that are not captured by shape, type, or weight update tests.
 
-- Reuse the already established setup for similar tests or establish a new reusable one.  **Rationale**: Ensures consistent test coverage and reduces effort to develop and maintain test.  Examples of reusable test setup can be found in `tests/conftest.py`.  This module contains reusable `pytest.fixtures` that have applicability across many tests.
+## What not to test
 
-- TorchTyping tests -- add typechecking based on torch input/outputs. **Rationale**: Allows for stronger typechecking of the form \[batch_size, dim1, dim2, ...\]. We can use the torch typing library to add these tests.
+It isn't necessary to test every function, a good rule is that if a function or method fulfills a requirement and will
+be called from outside the module, it should have a test.
 
-- Overfitting tests — ensure that a small ECD model is able to overfit on a small dataset. **Rationale**: Ensures that models are able to converge on reasonable targets and catches any unscoped issues that aren’t captured by shape/weight update tests.
+Things that don't need unit tests:
+
+- Constructors or properties. Test them only if they contain validations.
+- Configurations like constants, readonly fields, configs, etc.
+- Facades or wrappers around other frameworks or libraries.
+- Private methods
 
```

Comment:
Can we include a section on best practices? It could be useful to people who are new to software engineering.

```
## Best practices

There's lots of great advice on the web for writing good tests. Here are a few highlights [from Microsoft's recommendations](https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices) that we ascribe to:

- [Characteristics of a good unit test](https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#characteristics-of-a-good-unit-test)
- [Arranging unit tests](https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#arranging-your-tests)
- [Write minimally passing tests](https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#write-minimally-passing-tests)
- [Avoid logic in tests](https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#avoid-logic-in-tests)
- [Avoid multiple acts](https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#avoid-multiple-acts)
```

Reply:
Great idea!  Added.